[WIP] Make metrics containing different filters usable together, make metrics usable in expressions

lbrdnk opened the pull request:
Closes #25455 #30574

And probably also #13437.

Edited to be more descriptive.

### Description

This PR aims to solve multiple problems:
1. When metrics with different filters are used together, their filters are combined in root query, hence source data for particular aggregations is corrupted.
2. Metrics are not usable in expressions.

### Solution overview

Metrics are gathered from aggregations. Metrics are grouped per filter that is used in their definition. For every group `metrics-query` is created. Filter of metrics query is combined original query filter with filter from metrics definition. Values for metrics are computed in those queries. All of the metrics queries are then joined to the original query. Metrics clauses in aggregations of original query are swapped for fields coming from joined metrics queries. In case there is a aggregation containing only one metric and nothing else, it is swapped for a field. Those pending fields are then moved to the breakout. This step could potentially change the column order of a result. Hence resulting query is wrapped in ordering query, which sets the original order to fields. Also references in order-by are updated accordingly.

#### Combining filters

Combining filters guarantees that at least one row from joining query corresponds every element of metrics query that is being joined. That allows for use of left join operator.

#### Join condition
Breakout is shared between metrics-query and joining query. Every metric value coming from the metrics-query has its breakout combination. Metrics queries are joined on equal breakout combination values. That spreads out previously aggregated metric value across original rows, which when grouped in joining query, do not affect the already computed value of a metric.

#### Metrics and "groupped by"

Even though interface for metric modelling contains \"groupped by\" column, there is no groupping functionality  in place as per [flamber's comment](https://discourse.metabase.com/t/cannot-use-grouped-by-in-metrics/11339/2). Related issue can be found [here](https://github.com/metabase/metabase/issues/13167). Because of that, metric is expected to have no `:breakout` set and it is up to query that is using the metrics to set `:breakout` that will be common for all metrics contained in that query.

### How to verify

WIP

### Checklist

- [ ] Tests have been added/updated to cover changes in this PR

lbrdnk committed:
Update segment expansion

Segment expansion code is not bound to be used on `outer-query` only.
This will be taken advantage of during metrics' filter expansion.

lbrdnk committed:
Add option to disable prefixing joined field

Modify `annotate` middleware to provide option disabling addition
of join alias prefix to display name of field.

`expand-metrics` uses joins internally and this option makes it possible
to present user display names without internal join aliases in GUI.

lbrdnk committed:
[WIP] Draft to create draft PR

metamben commented:
I haven't had much time unfortunately, I've just read the description in the PR so far.

In general, the idea of running separate queries for different filters seems correct to me. Any group by in the metric should be taken into account the same way.

I don't quite understand how the separate queries would be joined to the original query though. What would the join condition be? I think the plan regarding the combination of the results should be refined/detailed a bit more.

I don't currently see any super easy ways to implement this, I think a fair amount of complexity is inherent in the problem.

As far as I know the plans for metrics 2.0 concern the relationship of metrics and models as well as the organization and handling of them, I don't expect a huge change related to their execution.

lbrdnk committed:
Add simple metric test

Assures that query with one metric containing just aggregation
returns expected results.

lbrdnk committed:
Add test for metric containing filter

lbrdnk committed:
Add test for metric containing segment

lbrdnk committed:
Thinking notes

lbrdnk committed:
Doc metrics and joins

lbrdnk committed:
Doc metrics and groupped by

lbrdnk committed:
Doc metrics and filter or segments

lbrdnk committed:
Doc solution

- groupping
- metrics query generation
- joining metrics query

lbrdnk committed:
Describe metrics query keys

lbrdnk committed:
Set warn-on-reflection, namespace contains interop

lbrdnk committed:
Thinking

lbrdnk committed:
Thinking

lbrdnk committed:
I was wrong with joins - doc

lbrdnk committed:
I was wrong with source query

Query contianing metrics should have source table corresponding to
metrics it is using.

lbrdnk committed:
Cleanup `expand-metrics*`

lbrdnk committed:
Whtiespace

lbrdnk committed:
WIP Test with expressions in breakout

lbrdnk committed:
Make it work with expressions

Expressions can be eg in breakout. Also factor out join condition code.

lbrdnk committed:
WIP refactor transform logic, without ordering now

lbrdnk committed:
Ordering rewrite mock

lbrdnk committed:
Whitespace, style

lbrdnk committed:
Ordering query works, next correct name prefixing?

lbrdnk committed:
Cleanup ordering query

lbrdnk committed:
Cleanup ::metric-id->field

lbrdnk committed:
Modify metric internal keys removal

Avoid redundant removals mostly. With this implementation, some keys,
namely field optoins, are leaved in place. Namespaced keys from query
are removed right after their last usage! This behavior is further
to be tested.

Food for thought. I could completely avoid use of namespaced keys
in query map and use them in field options, which is completely
legal (at least I think) with regards metadata computation.

lbrdnk committed:
Cleanup, remove now unused code

lbrdnk committed:
Condition code cleanup

lbrdnk committed:
Doc join condition code

lbrdnk committed:
Docstring `metrics-join`

lbrdnk committed:
Make swap functions respect metric id

When aggregation was to be transformed to field, and that field
was comming from deeper level nested metric, id from deeper metric was
used errorneosly. In case that current aggregation is from metric.
Now id is of "this aggregation" is swapped for id of deeper aggregation
and can correctly be provided upwards.

Code still in debug mode.

lbrdnk committed:
WIP Add recursively defined metric test

lbrdnk committed:
Refactor `provides`

lbrdnk committed:
Dead code removal

lbrdnk committed:
Update swapping logic

Fix use of `one-and-only-metric?` and respect metric id when swapping
for a field comming from deeper level metric definition. In case metric
is defined with use of other metric.

lbrdnk committed:
Doc

lbrdnk committed:
Use some? instead fo pos?

`pos?` is not compatible with nil.

lbrdnk committed:
Doc

lbrdnk committed:
Doc

lbrdnk committed:
Avoid nested partial, `swap-metric-clauses`

lbrdnk committed:
Update swap ag for field logic

lbrdnk committed:
Add solution description into ns docstr

lbrdnk committed:
Remove `segments` fn

Separate function for gathering segmetns from form is not necessary.

lbrdnk committed:
Refactor

lbrdnk committed:
Polish `adjust-aggregation-and-breakout`

lbrdnk committed:
doc

lbrdnk committed:
Doc `update-order-by-ag-refs`

lbrdnk committed:
Doc

lbrdnk committed:
Use more descriptive name - `move-ag-fields-to-breakout`

lbrdnk committed:
Simplify metric swapping code

lbrdnk committed:
Simplify `infer-ordered-clauses-for-fields` code, doc

lbrdnk committed:
Use subvec instead of drop

lbrdnk committed:
tmp

lbrdnk committed:
WIP Use expressions for fields

This way I can duplicate and rename fields comming from the metrics.
Will be explained later. Code is still in heavy WIP mode.

lbrdnk committed:
Expressions half refactored, now for the joins and provides part

lbrdnk committed:
WIP nested test passing, clean up the code next

lbrdnk committed:
WIP Testcases working after use of exprs

lbrdnk committed:
WIP ordered-clause-refs-for-fields

lbrdnk committed:
Make expand-macros ns readable, still debug mode.

lbrdnk committed:
Comment

lbrdnk committed:
Merge remote-tracking branch 'origin/master' into metrics

lbrdnk committed:
WIP Tests

lbrdnk committed:
WIP works apart from frontend scenario -- to be described in PR descr

lbrdnk commented:
Closing as per [this comment](https://github.com/metabase/metabase/issues/25455#issuecomment-1810338837).
